go/tools/builders/protoc.go

```diff
@@ -62,47 +74,77 @@ func run(args []string) error {
 	if err := cmd.Run(); err != nil {
 		return fmt.Errorf("error running protoc: %v", err)
 	}
-	notFound := []string{}
-	for _, src := range expected {
-		if _, err := os.Stat(src); os.IsNotExist(err) {
-			notFound = append(notFound, src)
+	// Build our file map, and test for existance
+	files := map[string]*FileInfo{}
+	byBase := map[string]*FileInfo{}
+	for _, path := range expected {
+		info := &FileInfo{
+			Path:     path,
+			Base:     filepath.Base(path),
+			Expected: true,
+			Unique:   true,
+		}
+		files[info.Path] = info
+		if byBase[info.Base] != nil {
+			info.Unique = false
+			byBase[info.Base].Unique = false
+		} else {
+			byBase[info.Base] = info
 		}
 	}
-	if len(notFound) > 0 {
-		missing := []string{}
-		unexpected := []string{}
-		filepath.Walk(".", func(path string, f os.FileInfo, err error) error {
-			if strings.HasSuffix(path, ".pb.go") {
-				wasExpected := false
-				matches := []string{}
-				base := filepath.Base(path)
-				for _, s := range expected {
-					if s == path {
-						wasExpected = true
-					}
-					if base == filepath.Base(s) {
-						matches = append(matches, s)
-					}
-				}
-				if !wasExpected {
-					if len(matches) != 1 {
-						unexpected = append(unexpected, path)
-					} else {
-						// Unambiguous mapping to expected output, so copy it
-						data, err := ioutil.ReadFile(path)
-						if err != nil {
-							return err
-						}
-						if err := ioutil.WriteFile(matches[0], data, 0644); err != nil {
-							return err
-						}
-					}
-				}
-			}
+	// Walk the generated files
+	filepath.Walk(".", func(path string, f os.FileInfo, err error) error {
+		if !strings.HasSuffix(path, ".pb.go") {
```

Comment:
Can't the suffix be something other than .pb.go? I thought the grpc gateway plugin generated something else.

Reply:
I have never run or seen the output of grpc-gateway, you may be right, who knows.
We can add an extra command line arg that passes in the expected extension from the compiler rule when we reach that bridge.

Reply:
Drive-by comment...  Yes it can be `.pb.gw.go`.  I'm not sure there is an overall generalization.

https://github.com/pubref/rules_protobuf/blob/master/grpc_gateway/BUILD#L8

Reply:
Okay, thanks!
I guess the real question is, do we even need to do this filtering, in theory there should not be any spare random files loafing around in this directory anyway... or maybe just .go is enough.